README

  Introduction
  ------------

mbutils - is a set of utilities and an object oriented library to work with
devices via communication protocol called Modbus. For more details on protocol
itself see: http://www.modbus.org. All code written in the Object Pascal and
compiled using FreePascal compiler (http://www.freepascal.org) with Lazarus
IDE. Please note, that library itself do not contain any visual components.


  Package contents
  ----------------

Package consists of two major parts and third party library:

  1. mbulib - the library itself. It uses an object oriented paradigm and was
designed to be easily extendible.

  2. mbutils - the set of utilities which can be used in debugging, deployment
and other activities related to Modbus devices.

  3. synapse - is an excellent library which hides all the details when working with
serial ports or TCP sockets. It is not strictly required to use the same version
of synapse which is included with this package, but one have to verify that both
are compatible.

It is assumed, that users of this package have enough knowledge of Modbus protocol
because later we'll rarely repeat official documentation.


  Installation
  ------------

Unlike usual programs, this library does not need an installation step. Just
unpack it somewhere (say under src/pas/mbutils name). Supplied utilities can be
compiled using Lazarus IDE. When you we'll need to use mbulib in your code, it
is recommended to copy original files into the project directory. This may
sound a bit strange, but in practise works well because each project may have
its own version of the library and in the case of API changes you'll not need to
revise all the projects to get them compiled. But in general, you may work with
the library as you wish.

Advanced users may wish to use a direct checkout from the subversion repository.


  Code example
  ------------

  Although mbulib was tailored to suit the needs of multi-threaded programs,
here we'll demonstrate a blocked example:

var
  mbc : TMBRTUMaster;
  q   : TMBQuery;
begin
  mbc := TMBRTUMaster.Create;
  mbc.Baudrate := 9600;
  q := TMBQReadHoldingRegisters.Create(1);
  q.rq.StartReg := 200;
  q.rq.RegCnt := 5;
  mbc.Connect;
  mbc.SubmitQuery(q);
  q.WaitForCompletion;
  writeln('Reg 200=', q.rp.RegU16BE[0]);
  q.Free;
  mbc.Free;
end;

The code are quite straight forward and only minor comments are needed. In the
first two steps we create and setup the RTU master (client) and Modbus
request/reply pair called 'query'.  Then with 'Connect' call master tries to
occupy serial device and 'SubmitQuery' sends a prepared request. At this point
program may continue with other tasks, but instead it merely waits when
processing of request will be done. The 'GetCompletedQuery' call flushes
receive queue of master and now we ready to display received data. Of course,
this sample completely omits error checking, but the goal was to provide the
main code path.

It is highly recommended to look into the code of supplied utilities - they
are simple enough to understand and may serve as good starting point.


  Bug reporting
  -------------

All programs have bugs, this is inevitable. So, if you think that there is a
bug, then drop me a note, preferably with working example, or even more
preferable write a patch because mbulib are very transparent and it shouldn't
be too hard to debug it.


  Licensing
  ---------

The library itself released under LGPL v2 license. All utilities released under
GPL v2 license. If in doubt, look at first lines of particular source file.


trunk/HISTORY
24.09.2013	Add example of Modbus proxy utility

22.11.2010	Added support for signed big-endian values.

17.11.2010	mbuquery utility of version 1.3 made available

16.11.2010	initial code check-in


mbulib/mbutcp.pas
{$REGION 'license'}
{
 * This file is a part of mbulib.
 * Copyright (C) 2008-2010 by Boris Popov <borisxm@gmail.com>.
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
}
{$ENDREGION}
unit mbutcp;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils,
  blcksock, mbubase;

type
   TMBIOState = (mbiosIdle, mbiosTx, mbiosRx, mbiosSilence);
 { TMBTCPClient }

  TMBTCPClient = class(TMBClient)
  private
    function GetQWait(AIndex: integer): TMBquery;
  private
    procedure SendQuery(AQ: TMBQuery);
    procedure Receive;
    procedure PurgeOld;
    function LookupByQID(AQID: cardinal): TMBQuery;
    procedure PutOnWait(AQ: TMBQuery);
    procedure TakeFromWait(AQ: TMBQuery);

    property QWait[AIndex: integer]: TMBquery read GetQWait;
  protected
    FTargetHost		: string;
    FTargetPort		: string;
    FIPInterface	: string;
    FSock		: TTCPBlockSocket;
    FState		: TMBIOState;
    FMaxPendingQueries	: cardinal;	// Maximum number of queries which will be enqueued to server
    FWaitCnt		: cardinal;	// Number of queries on hold
    FQWait		: TList;	// List of queries wating for respose
    FTxBuffer		: array of byte;
    FRxBuffer		: array of byte;
    FBufferSize		: integer;	// Maximum length of Tx and Rx buffers
    FReplyTimeout	: cardinal;	// Timeout for wating a reply (1s)
    FRxTimeout		: cardinal;	// Intercharacter timeout (1.5 * ct)
    FSilencePeriod	: cardinal;	// Time between exchanges on the bus (3.5 * ct)

    procedure Execute;override;
    procedure Initialise;
    procedure Finalise;override;
  public
    constructor Create;
    destructor Destroy;override;
    function Connect:boolean;override;
    procedure Disconnect;override;

    property Sock: TTCPBlockSocket read FSock;
    property TargetHost: string read FTargetHost write FTargetHost;
    property TargetPort: string read FTargetPort write FTargetPort;
    property ReplyTimeout: cardinal read FReplyTimeout write FReplyTimeout;
    property RxTimeout: cardinal read FRxTimeout;
    property SilencePeriod: cardinal read FSilencePeriod;
  end;

implementation

uses
  synautil;

{ TMBTCPClient }

constructor TMBTCPClient.Create;
begin
  inherited Create;
  FBufferSize := MB_RTU_PKT_LEN_MAX + 7;
  FMaxPendingQueries := 10;
  SetLength(FRxBuffer, FBufferSize);
  SetLength(FTxBuffer, FBufferSize);
  FTargetHost := 'localhost';
  FTargetPort := '502';
  FIPInterface := cAnyHost;
  FReplyTimeout := 1000;

  FQWait := TList.Create;
  FSock := TTCPBlockSocket.Create;
  FSock.Owner := self;
end;

destructor TMBTCPClient.Destroy;
begin
  Disconnect;
  FreeAndNil(FQWait);
  FreeAndNil(FSock);
  inherited Destroy;
end;

function TMBTCPClient.Connect: boolean;
begin
  result := Connected;
  if result then
    exit;

  FSock.Bind(FIPInterface, cAnyPort);
  if FSock.LastError <> 0 then
    exit;
  FSock.Connect(FTargetHost, FTargetPort);
  try
    MustDie := false;
    Resume;
  except
    FSock.CloseSocket;
  end;
  FConnected := true;
end;

procedure TMBTCPClient.Disconnect;
var
  i	: integer;
  q	: TMBQuery;
begin
  if not Connected then
    exit;
  try
    FSock.CloseSocket;
    FreeOnTerminate := false;
    MustDie := true;
    FSubmitEvent.SetEvent;
    while not Terminated do begin
//	  CheckSynchronize(0);
      Sleep(1);
    end;
    for i := 0 to pred(FQWait.Count) do begin
      q := QWait[i];
      if q = nil then
        continue;
      q.Complete(mbeCommFailed);
    end;
  finally
    FConnected := false;
  end;
end;

function TMBTCPClient.LookupByQID(AQID: cardinal): TMBQuery;
var
  i	: integer;
begin
  for i := 0 to pred(FQWait.Count) do begin
    result := QWait[i];
    if result = nil then
      continue;
    if result.QId = AQid then
      exit;
  end;
  result := nil;
end;

procedure TMBTCPClient.PutOnWait(AQ: TMBQuery);
var
  i	: integer;
begin
  i := FQWait.IndexOf(nil);
  if i <> -1 then
    FQWait[i] := AQ
  else
    FQWait.Add(AQ);
  inc(FWaitCnt);
end;

procedure TMBTCPClient.TakeFromWait(AQ: TMBQuery);
begin
  dec(FWaitCnt);
  FQWait[FQWait.IndexOf(AQ)] := nil;
end;

function TMBTCPClient.GetQWait(AIndex: integer): TMBquery;
begin
  result := TMBQuery(FQWait[AIndex]);
end;

procedure TMBTCPClient.SendQuery(AQ: TMBQuery);
var
  rqlen, qid	: word;
begin
  rqlen := AQ.Rq.Len;
  inc(rqlen, 1);		// slaveid
  qid := NextSeq;
  AQ.QId := qid;
  // Fill MBAP
  mbu_PutU16BE(FTxBuffer[0], qid);	// TID
  mbu_PutU16BE(FTxBuffer[2], 0);		// Protocol ID
  mbu_PutU16BE(FTxBuffer[4], rqlen);	// Length
  FTxBuffer[6] := AQ.SlaveAddr;		// UnitId
  AQ.Rq.GetPDU(FTxBuffer[7]);
  inc(rqlen, 6);
  FSock.Purge;
  if FSock.SendBuffer(@FTxBuffer[0], rqlen) = rqlen then begin
    AQ.StartTime := synautil.GetTick;
    PutOnWait(AQ);
  end else begin
    aq.Complete(mbeTxFailed);
  end;
end;

type
  TMBAP = packed record
    TID		: word;
    ProtoID	: word;
    PktLength	: word;
    SlaveId	: byte;
  end;

procedure TMBTCPClient.Receive;
var
  rp	: TMBPDU;
  nRead, toRead: integer;
  mbap	: TMBAP;
  q	: TMBQuery;
begin
  nread := FSock.RecvBufferEx(@mbap, sizeof(mbap), 1);
  if nread <> sizeof(mbap) then
    exit;

  q := LookupByQID(mbu_GetU16BE(mbap.TID));
  toRead := mbu_GetU16BE(mbap.PktLength);
  if (q = nil) or (toRead > MB_PDU_LEN_MAX) then begin
    FSock.RecvBufferEx(@FRxBuffer[0], MB_RTU_PKT_LEN_MAX, 1);	// try to discard the rest of packet
    exit;
  end;

  nread := FSock.RecvBufferEx(@FRxBuffer[0], toRead, 1);

  rp := q.rp;

  rp.SetPDU(FRxBuffer[0], toRead - 1);
  TakeFromWait(q);
  q.Complete(rp.ParseReply);
  FRxQueue.Push(q);
end;

procedure TMBTCPClient.PurgeOld;
var
  i	: integer;
  q	: TMBQuery;
  tnow	: longword;
begin
  tnow := synautil.GetTick;
  for i := 0 to pred(FQWait.Count) do begin
    q := QWait[i];
    if q = nil then
      continue;
    if synautil.TickDelta(q.StartTime, tnow) > FReplyTimeout then begin
      q.Complete(mbeTimeout);
      TakeFromwait(q);
    end;
  end;
end;

procedure TMBTCPClient.Execute;
var
  rq	: TMBQuery;
begin
  try
    Initialise;
    while not MustDie do begin
      try
        // Process incoming data
        if FSock.CanReadEx(1) then begin
          Receive;
        end;

        // Try to send data
        if FWaitCnt < FMaxPendingQueries then begin
          rq := FTxQueue.Pop as TMBQuery;
          if (rq <> nil) then begin
            SendQuery(rq);
          end;
        end;

        // Look through, and discard timedout queries
        PurgeOld;
      finally
      end;
    end;
  finally
    Terminate;
    Finalise;
  end;
end;

procedure TMBTCPClient.Initialise;
begin

end;

procedure TMBTCPClient.Finalise;
begin
  inherited Finalise;
end;

end.



mbulib/blogger.pas
{$REGION 'license'}
{
 * Multistream message logger.
 * Copyright (C) 2001-2003 by Boris Popov <borisxm@gmail.com>.
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
}
{$ENDREGION}
unit blogger;

interface

uses
  SysUtils, Classes, bqueue;

type
  TBLogLevel = (bllError, bllWarning, bllNotice, bllInfo, bllDebug);

  { TBLogConsumer }

  TBLogConsumer = class
  private
    FLevel	: TBLogLevel;
  public
    constructor Create;
    procedure Write(ALevel: TBLogLevel; const AMsg: string);virtual;abstract;

    property Level: TBLogLevel read FLevel write FLevel;
  end;

  { TBLogToConsole }

  TBLogToConsole = class(TBLogConsumer)
  public
    procedure Write(ALevel: TBLogLevel; const AMsg: string);override;
  end;

  { TBLogToFile }

  TBLogToFile = class(TBLogConsumer)
  private
    FFileName	: string;
    FHandle	: integer;
    FMode	: word;
    FFilePos	: longint;
    FMaxSIze	: longint;
  public
    constructor Create(const AFileName: string; ATruncate: boolean = false);
    destructor Destroy;override;
    procedure Write(ALevel: TBLogLevel; const AMsg: string);override;
  end;

  { TBLogger }

  TBLogger = class
  private
    FConsumers	: TList;
    FFacility	: string;
    FProcName	: string;
  public
    constructor Create;
    destructor Destroy;override;
    procedure Add(ALevel: TBLogLevel; const AMsg: string);overload;
    procedure Add(ALevel: TBLogLevel; const AMsg: string; AFmt: array of const);overload;
    procedure RegisterConsumer(ALC: TBLogConsumer);

    property Facility: string read FFacility write FFacility;
    property ProcName: string read FProcName write FProcName;
  end;

const
  Log: TBLogger = nil;

implementation

{ TBLogger }

constructor TBLogger.Create;
begin
  inherited;
  FConsumers := TList.Create;
  FFacility := '';
  FProcName := '';
end;

destructor TBLogger.Destroy;
var
  i	: integer;
begin
  for i := 0 to pred(FConsumers.Count) do
    if FConsumers[i] <> nil then
      TBLogConsumer(FConsumers[i]).Free;
  FreeAndNil(FConsumers);
  inherited Destroy;
end;

procedure TBLogger.Add(ALevel: TBLogLevel; const AMsg: string; AFmt: array of const);
begin
  if FConsumers.Count = 0 then
    exit;
  Add(ALevel, Format(AMsg, AFmt));
end;

procedure TBLogger.Add(ALevel: TBLogLevel; const AMsg: string);
var
  i	: integer;
  lc	: TBLogConsumer;
  msg	: string;
begin
  if FConsumers.Count = 0 then
    exit;
  msg := Format('%s %s[%u]: %s', [FormatDateTime('m d hh:nn:ss', Now), FProcName, system.GetProcessID, AMsg]);

  for i := 0 to pred(FConsumers.Count) do begin
    lc := TBLogConsumer(FConsumers[i]);
    if lc = nil then
      continue;
    lc.Write(ALevel, msg);
  end;
end;

procedure TBLogger.RegisterConsumer(ALC: TBLogConsumer);
begin
  FConsumers.Add(ALC);
end;

{ TBLogConsumer }

constructor TBLogConsumer.Create;
begin
  FLevel := bllError;
end;


{ TBLogToConsole }

procedure TBLogToConsole.Write(ALevel: TBLogLevel; const AMsg: string);
begin
  if ALevel >= FLevel then
    exit;
  writeln(stderr, AMsg);
end;

{ TBLogToFile }

constructor TBLogToFile.Create(const AFileName: string; ATruncate: boolean);
begin
  inherited Create;
  FFileName := AFileName;

  if ATruncate or not FileExists(FFileName) then begin
    FHandle := FileCreate(FFileName, fmCreate);
    if FHandle < 0 then
      raise Exception.CreateFmt('%s: Unable to create file %s.', [self.ClassName, FFileName]);
    FileClose(FHandle);
  end;

  FHandle := FileOpen(FFileName, fmOpenReadWrite or fmShareDenyWrite);
  if FHandle < 0 then
      raise Exception.CreateFmt('%s: Unable to open file %s.', [self.ClassName, FFileName]);
  FFilePos := FileSeek(FHandle, 0, fsFromEnd);
end;

destructor TBLogToFile.Destroy;
begin
  if FHandle >= 0 then
    FileClose(FHandle);
  inherited Destroy;
end;

procedure TBLogToFile.Write(ALevel: TBLogLevel; const AMsg: string);
const
  crlf:string = #10;
var
  nwr	: longint;
begin
  if ALevel >= FLevel then
    exit;
  if (FMaxSize <> 0) and (FFilePos > FMaxSize) then
    exit;	// TODO: rotate file?
  nwr := FileWrite(FHandle, AMsg[1], Length(AMsg));
  if nwr < 0 then begin
    writeln(stderr, Format('%s: Unable to write file %s.', [ClassName, FFileName]));
    exit;
  end;
  FileWrite(FHandle, crlf, 1);
  inc(FFilePos, nwr + 1);
end;

initialization
  log := TBLogger.Create;

finalization
  FreeAndNil(log);
end.



trunk/mbulib/cmdlparse.pas
{$REGION 'license'}


{
 * Simple, yet powerful and extensible command line parser.
 * Copyright (C) 2001-2010 by Boris Popov <borisxm@gmail.com>.
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
}
{$ENDREGION}

{:
  @abstract(Simple, yet powerful and extensible command line parser.)
}
unit cmdlparse;

interface
{$IFDEF FPC}
  {$mode objfpc}{$H+}
{$ENDIF}

uses
  SysUtils, Classes;

const
  {: Used to inidicate that a 'novalue' condition for integer fields. }
  TCLP_NoIntegerValue	= $0ABBCCDD;

type
  {: All exceptions specific to parser should be of this type. }
  ECommandLineParser = class(Exception);

  { forward declarations }
  TCommandLineParser = class;

  { TCommandLineItem }

  {: @abstract(Parent class for all types of command line switches. }
  TCommandLineItem = class
  protected
    FLongName	: string;		// 'help', 'verbose', etc
    FDesc	: string;		// 'Print this text and exit'
    FArgCount	: cardinal;		// Number of required arguments
    FShortName	: char;			// 'h', 'v'
    FPresent	: boolean;
    FRequired	: boolean;		// option must be present

    procedure ParseArgument(const Arg: string);virtual;
    function ArgumentType:string;virtual;
  public
    constructor Create(AOwner: TCommandLineParser; const ALongName: string; AShortName: char; const ADesc: string);
    function ExplainDefaults:string;virtual;
    function Explanation:string;

    property LongName: string read FLongName;
    property ShortName: char read FShortName;
    property Desc: string read FDesc;
    property Present: boolean read FPresent;
    property ArgCount: cardinal read FArgCount;
    property Required: boolean read FRequired write FRequired;
  end;

  {: @abstract(Parameterless options like -h or -v.) }
  TCommandLineFlag = class(TCommandLineItem)
  protected
  end;

  { TCommandLineString }

  {: @abstract(Option which requires a string as argument.) }
  TCommandLineString = class(TCommandLineItem)
  protected
    FValue		: string;
    FDefault		: string;
    FValidValues	: TStrings;

    procedure ParseArgument(const Arg: string);override;
    function ArgumentType:string;override;
  public
    constructor Create(AOwner: TCommandLineParser; const ALongName: string; AShortName: char;
      const ADesc: string; ADefault: string = '');
    constructor Create(AOwner: TCommandLineParser; const ALongName: string; AShortName: char;
      const ADesc: string; ADefault: string; const AValidValues:array of string);
    destructor Destroy;override;
    function ExplainDefaults:string;override;

    property Value: string read FValue;
  end;

  { TCommandLineInteger }

  {: @abstract(Option which requires a an integer as argument.) }

  TCommandLineInteger = class(TCommandLineItem)
  protected
    FValue	: integer;
    FDefault	: integer;
    FMin	: integer;
    FMax	: integer;
    procedure ParseArgument(const Arg: string);override;
    function ArgumentType:string;override;
  public
    constructor Create(AOwner: TCommandLineParser; const ALongName: string; AShortName: char;
      const ADesc: string; ADefault: integer = 0;
      AMin: integer = TCLP_NoIntegerValue; AMax: integer = TCLP_NoIntegerValue);
    function ExplainDefaults:string;override;

    property Value: integer read FValue;
  end;

  { TCommandLineParser }

  {: @abstract(The parser by itself.) }
  TCommandLineParser = class
  private
    function GetOption(const AIndex: cardinal): TCommandLineItem;
  protected
    FProgramName	: string;
    FFullProgramPath	: string;
    FOptionsList	: TList;
    FArgumentsList	: TStrings;
    FShortOptionPrefix	: char;
    FParseDone		: boolean;

    function LookupOption(const AName: string): TCommandLineItem;
    function LookupOptionByShortName(AName: char): TCommandLineItem;
  public
    constructor Create;
    destructor Destroy;override;
    procedure Add(AOption: TCommandLineItem);
    function OptionsCount: cardinal;
    function OptionsHelp:string;
    function OptionPresent(const AName: string): boolean;
    {: Parse command line. }
    procedure Parse;overload;
    {: Parse provided string. }
    procedure Parse(const cline:string);overload;
    {: Parse provided string. }
    procedure Parse(cp: pchar);overload;
    class function GetToken(cp: pchar; var cnext: pchar): string;

    property Options[const AIndex: cardinal]: TCommandLineItem read GetOption;
    property ShortOptionPrefix: char read FShortOptionPrefix write FShortOptionPrefix;
    property Arguments: TStrings read FArgumentsList;
    property ProgramName: string read FProgramName write FProgramName;
    property FullProgramPath: string read FFullProgramPath;
  end;

const
  {: Width of options column in the help messages. }
  OptionsColumnWidth: integer = 25;

implementation

{ TCommandLineItem }
constructor TCommandLineItem.Create(AOwner: TCommandLineParser; const ALongName: string; AShortName: char; const ADesc: string);
begin
  FLongName := ALongName;
  FShortName := AShortName;
  FDesc := ADesc;
  FPresent := false;
  if AOwner <> NIL then
    AOwner.Add(self);
end;

function TCommandLineItem.ExplainDefaults: string;
begin
  result := '';
end;

function TCommandLineItem.Explanation: string;
var
  i	: integer;
  argtype : string;
begin
  if FShortName <> #0 then
    result := '-' + FShortName
  else
    result := '';

  if FLongName[1] <> '-' then begin
    if result <> '' then
      AppendStr(result, ', ');
    AppendStr(result, '--' + FLongName);
  end;

  argtype := ArgumentType;
  if argtype <> '' then
    AppendStr(result, ' <' + argtype + '>');

  i := OptionsColumnWidth - Length(result);
  while i > 0 do begin
    AppendStr(result, ' ');
    dec(i);
  end;
  AppendStr(result, FDesc);
  argtype := ExplainDefaults;
  if argtype <> '' then
    AppendStr(result, ' (' + argtype + ')');
end;

procedure TCommandLineItem.ParseArgument(const Arg: string);
begin
  // do nothing, or better raise an error
end;

function TCommandLineItem.ArgumentType: string;
begin
  result := '';
end;


{ TCommandLineString }

constructor TCommandLineString.Create(AOwner: TCommandLineParser;
  const ALongName: string; AShortName: char; const ADesc: string;
  ADefault: string);
begin
  inherited Create(AOwner, ALongName, AShortName, ADesc);
  FValue := ADefault;
  FDefault := ADefault;
  FArgCount := 1;
  FValidValues := nil;
end;

constructor TCommandLineString.Create(AOwner: TCommandLineParser; const ALongName: string;
   AShortName: char; const ADesc: string; ADefault: string; const AValidValues:array of string);
var
  i : integer;
begin
  Create(AOwner, ALongName, AShortName, ADesc, ADefault);

  FValidValues := TStringList.Create;
  for i := low(AValidValues) to high(AValidValues) do
    FValidValues.Add(AValidValues[i]);
end;

destructor TCommandLineString.Destroy;
begin
  FreeAndNil(FValidValues);
  inherited Destroy;
end;

function TCommandLineString.ExplainDefaults: string;
var
  i : integer;
begin
  if FValidValues <> nil then begin
    result := '';
    for i := 0 to pred(FValidValues.Count) do begin
       if result <> '' then
         if i <> (FValidValues.Count - 1) then
           AppendStr(result, ', ')
         else
           AppendStr(result, ' or ');
       AppendStr(result, '''' + FValidValues[i] + '''');
    end;
  end else
    result := '';

  if FDefault <> '' then begin
    if result <> '' then
      AppendStr(result, ', ');
    AppendStr(result, '''' + FDefault + ''' is default');
  end;
end;

procedure TCommandLineString.ParseArgument(const Arg: string);
begin
  FValue := Arg;
  if FValidValues = NIL then
    exit;

  if FValidValues.IndexOf(Arg) <> -1 then
    exit;
  raise ECommandLineParser.CreateFmt('Value ''%s'' is not allowed here', [Arg]);
end;

function TCommandLineString.ArgumentType: string;
begin
  result := 'string';
end;

{ TCommandLineInteger }

constructor TCommandLineInteger.Create(AOwner: TCommandLineParser; const ALongName: string;
  AShortName: char; const ADesc: string; ADefault: integer;
  AMin: integer = TCLP_NoIntegerValue; AMax: integer = TCLP_NoIntegerValue);
begin
  inherited Create(AOwner, ALongName, AShortName, ADesc);
  FValue := ADefault;
  FDefault := ADefault;
  FMin := AMin;
  FMax := AMax;
  FArgCount := 1;
end;

function TCommandLineInteger.ExplainDefaults: string;
begin
  case FMax - FMin of
    0: result := '';
    1: result := Format('%d or %d', [FMin, FMax]);
    2: result := Format('%d, %d or %d', [FMin, FMax - 1, FMax]);
  else
    result := Format('%d..%d', [FMin, FMax]);
  end;
  if (result = '') and (FDefault = TCLP_NoIntegerValue) then
    exit;
  if FDefault <> TCLP_NoIntegerValue then begin
    if result <> '' then
      AppendStr(result, ', ');
    AppendStr(result, IntToStr(FDefault) + ' is default');
  end;
end;

procedure TCommandLineInteger.ParseArgument(const Arg: string);
begin
  FValue := StrToInt(Arg);
  if (FMin <> TCLP_NoIntegerValue) and (FValue < FMin) then
    raise ECommandLineParser.CreateFmt('Value %d are less than %d', [FValue, FMin]);
  if (FMin <> TCLP_NoIntegerValue) and (FValue > FMax) then
    raise ECommandLineParser.CreateFmt('Value %d are greater than %d', [FValue, FMax]);
end;

function TCommandLineInteger.ArgumentType: string;
begin
  result := 'number';
end;

{ TCommandLineParser }

constructor TCommandLineParser.Create;
begin
  FOptionsList := TList.Create;
  FArgumentsList := TStringList.Create;
  FShortOptionPrefix := '-';
  FParseDone := false;
end;

destructor TCommandLineParser.Destroy;
var
  i	: integer;
begin
  for i := 0 to pred(OptionsCount) do
    Options[i].Free;
  FreeAndNil(FOptionsList);
  FreeAndNil(FArgumentsList);
  inherited;
end;

procedure TCommandLineParser.Add(AOption: TCommandLineItem);
begin
  FOptionsList.Add(AOption);
end;

function TCommandLineParser.OptionsCount: cardinal;
begin
  result := FOptionsList.Count;
end;

function TCommandLineParser.OptionsHelp: string;
var
  i: cardinal;
begin
  result := '';
  for i := 0 to pred(OptionsCount) do begin
    AppendStr(result, Options[i].Explanation);
{$ifdef Win32}
    AppendStr(result, #13#10);
{$else}
    AppendStr(result, #10);
{$endif}
  end;
end;

function TCommandLineParser.OptionPresent(const AName: string): boolean;
var
  opt	: TCommandLineItem;
begin
  opt := LookupOption(AName);
  if opt = NIL then
    raise ECommandLineParser.CreateFmt('Asking for undeclared option %s', [AName]);
  result := opt.FPresent;
end;

class function TCommandLineParser.GetToken(cp: pchar; var cnext: pchar):string;
const
  SPaces = [#9, #10, #13, #32];
var
  sp : pchar;
  st : char;
begin
  // skip leading spaces
  while cp[0] in Spaces do
    inc(cp);

  // determine quotation character
  if (cp[0] = '''') or (cp[0] = '"') then begin
    st := cp[0];
    inc(cp);
  end else
    st := #0;

  sp := cp;
  // look through token
  while not (cp[0] in Spaces + [st]) do
    inc(cp);
  if (st <> #0) and (cp[0] <> st) then
    raise ECommandLineParser.Create('Unterminated quoted string');
  result := Copy(sp, 1, cp - sp);
  if st <> #0 then
    inc(cp);
  cnext := cp;
end;

procedure TCommandLineParser.Parse;
begin
  Parse(cmdline);
end;

procedure TCommandLineParser.Parse(cp: pchar);
  procedure ParseError(const AMsg: string);
  begin
    raise ECommandLineParser.CreateFmt('Error: %s at %s', [AMsg, Copy(string(cp), 1, 5)]);
  end;

  procedure skipSpaces;
  begin
    while cp[0] in [' ', #9, #10, #13] do
      inc(cp);
  end;
var
  opt : TCommandLineItem;
  optname: string;
  optidx: cardinal;
  optend: pchar;
  i	: integer;
begin
  for optidx := 0 to pred(OptionsCount) do
    Options[optidx].FPresent := false;

  skipSpaces;
//  if (cp[0] = '/') or ((cp[0] <> #0) and (cp[1] = ':')) then begin
    FFullProgramPath := GetToken(cp, cp);
//  end;

  skipSpaces;
  while cp[0] <> #0 do begin
    if cp[0] <> FShortOptionPrefix then
      break;	// First non option
    inc(cp);
    if cp[0] = '-' then begin	// long option
      inc(cp);
      if cp[0] in [#0, ' ', #9] then // '--' - the end of the options list
        break;
      optend := cp;
      // Skip to the end of option name
      while not (optend[0] in [#0, ' ', #9, '=']) do
        inc(optend);
      optname := Copy(cp, 1, optend - cp);
      opt := LookupOption(optname);
      if opt = NIL then
        ParseError('Unknown option --' + optname);
      cp := optend;
      if (opt.FArgCount = 0) and (cp[0] = '=') then
        ParseError('Option may not have an argument');
      opt.FPresent := true;
      if opt.FArgCount = 0 then begin
        opt := NIL;
      end else begin
        inc(cp);
        skipSpaces;
        opt.ParseArgument(GetToken(cp, cp));
      end;
      skipSpaces;
      continue;
    end;

    while true do begin // short options list
      opt := LookupOptionByShortName(cp[0]);
      if opt = NIL then
        ParseError('Unknown option');
      opt.FPresent := true;
      inc(cp);
      if opt.FArgCount <> 0 then begin
        skipSpaces;
        opt.ParseArgument(GetToken(cp, cp));
        break;
      end;
      if cp[0] in [#0, #9, ' '] then
        break;
    end;
    skipSpaces;
  end; // while in options

  for i := 0 to pred(OptionsCount) do begin
    opt := Options[i];
    if opt.Required and not opt.Present then
      raise ECommandLineParser.CreateFmt('Option ''%s'' must be present.', [opt.Desc]);
  end;

  while true do begin
    skipSpaces;
    if cp[0] = #0 then
      break;
    FArgumentsList.Add(GetToken(cp, cp));
  end;
  FParseDone := true;
end;

procedure TCommandLineParser.Parse(const cline: string);
begin
  Parse(pchar(cline));
end;

function TCommandLineParser.GetOption(const AIndex: cardinal): TCommandLineItem;
begin
  result := TCommandLineItem(FOptionsList[AIndex]);
end;

function TCommandLineParser.LookupOption(const AName: string): TCommandLineItem;
var
  i	: cardinal;
begin
  for i := 0 to pred(OptionsCount) do begin
    result := Options[i];
    if result.LongName = AName then
      exit;
  end;
  result := NIL;
end;

function TCommandLineParser.LookupOptionByShortName(AName: char): TCommandLineItem;
var
  i	: cardinal;
begin
  for i := 0 to pred(OptionsCount) do begin
    result := Options[i];
    if result.FShortName = AName then
      exit;
  end;
  result := NIL;
end;

end.

